hawk/src/main/java/com/orhanobut/hawk/Hawk.java

```diff
@@ -111,20 +102,90 @@ public static void init(Context context, String password, LogLevel logLevel) {
      * @return true if put is successful
      */
     public static <T> boolean put(String key, List<T> list) {
+        return put(key, null, list, true, true) != null;
+    }
+
+    /*
+     * A helper put method.
+     *
+     * @param key          the key
+     * @param value        the value if isList is false, otherwise null
+     * @param list         the list, if isList is true, otherwise null
+     * @param isList       determines if a list instance should be present
+     * @param addToStorage determines if the keyval should be immediately added to storage
+     */
+    private static <T> String put(String key, T value, List<T> list, boolean isList,
+                                  boolean addToStorage) {
+        ensureKeyIsValid(key);
+        if (isList) {
+            ensureListIsValid(list);
+        } else {
+            ensureValueIsValid(value);
+        }
+
+        String cipherText;
+        if (isList) {
+            cipherText = encoder.encode(list);
+        } else {
+            cipherText = encoder.encode(value);
+        }
+
+        if (cipherText == null) {
+            return null;
+        }
+
+        String fullText;
+        if (isList) {
+            fullText = DataUtil.addType(cipherText, list.get(0).getClass(), true);
+        } else {
+            fullText = DataUtil.addType(cipherText, value.getClass(), false);
+        }
+
+        boolean successful = true;
+        if (addToStorage) {
+            successful = storage.put(key, fullText);
+        }
+
+        return successful ? fullText : null;
+    }
+
+    private static <T> void ensureListIsValid(List<T> list) {
```

Comment:
Let's keep the current structure as it is for the validation. 

About the empty list validation, It should be IllegalStateException. The reason is that the argument type is actually correct but the state is wrong. IllegalArgumentException is misleading